x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,526 @@
+        NumberFieldMapper minorVersionMapper,
+        NumberFieldMapper patchVersionMapper
+    ) {
+        super(simpleName, fieldType, mappedFieldType, multiFields, copyTo);
+        this.nullValue = nullValue;
+        this.prereleaseSubField = preReleaseMapper;
+        this.majorVersionSubField = majorVersionMapper;
+        this.minorVersionSubField = minorVersionMapper;
+        this.patchVersionSubField = patchVersionMapper;
+    }
+
+    @Override
+    public VersionStringFieldType fieldType() {
+        return (VersionStringFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected VersionStringFieldMapper clone() {
+        return (VersionStringFieldMapper) super.clone();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        String versionString;
+        if (context.externalValueSet()) {
+            versionString = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                versionString = nullValue;
+            } else {
+                versionString = parser.textOrNull();
+            }
+        }
+
+        if (versionString == null) {
+            return;
+        }
+
+        EncodedVersion encoding = encodeVersion(versionString);
+        BytesRef encodedVersion = encoding.bytesRef;
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
+            Field field = new Field(fieldType().name(), encodedVersion, fieldType);
+            context.doc().add(field);
+            // encode the first 16 bytes as points for efficient range query
+            byte[] first16bytes = Arrays.copyOfRange(encodedVersion.bytes, encodedVersion.offset, 16);
+            context.doc().add(new BinaryPoint(fieldType().name(), first16bytes));
+        }
+        context.doc().add(new SortedSetDocValuesField(fieldType().name(), encodedVersion));
+
+        // add additional information extracted from version string
+        context.doc().add(new Field(prereleaseSubField.name(), encoding.isPreRelease ? "T" : "F", BooleanFieldMapper.Defaults.FIELD_TYPE));
+        context.doc().add(new SortedNumericDocValuesField(prereleaseSubField.name(), encoding.isPreRelease ? 1 : 0));
+
+        addVersionPartSubfield(context, majorVersionSubField.name(), encoding.major);
+        addVersionPartSubfield(context, minorVersionSubField.name(), encoding.minor);
+        addVersionPartSubfield(context, patchVersionSubField.name(), encoding.patch);
+    }
+
+    private void addVersionPartSubfield(ParseContext context, String fieldName, Integer versionPart) {
+        if (versionPart != null) {
+            context.doc().addAll(NumberType.INTEGER.createFields(fieldName, versionPart, true, true, false));
```

Comment:
This is not aimed at making the range queries faster like in indexing the 16 byte prefix as point. I added these fields because the telemetry team have the use case that they want to filter and aggregate on mayor / minor version (maybe patch), e.g. with histograms or terms aggregations. Think along the lines of "give me a breakdown of all mayor/minor/patch" versions running in my cluster. That information would otherwise have to be parsed out again on aggregations e.g. by scripts, so adding these field will make these kind of aggregations faster. I was considering having to switch these additional subfields on with a type parameter so its not done by default.

Reply:
Ok thanks for explaining. Although I think we should discuss more alternatives to handle these requirements. Extracting the major/minor/patch shouldn't be considered costly in aggregations for instance. We also have plenty of ways to make the decoding faster if needed.
Querying all `0` minor should also  be possible with the full version so I'd prefer to find an alternative that doesn't require 3 additional fields.